### ISSUE TITLE
ComposerProjectTemplatesTest.php doesn't work

### ISSUE THREAD

**Issue by joachim-n** [joachim-n/drupal-core-development-project#10]
Core tests should pass with this template.

However, the core tests that cover the Composer project templates in core fail when run in this template:

> vendor/bin/phpunit web/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php --filter=testTemplateCreateProject@recommended-project

```

1) Drupal\BuildTests\Composer\Template\ComposerProjectTemplatesTest::testTemplateCreateProject with data set "recommended-project" ('drupal/recommended-project', 'composer/Template/RecommendedProject', '/web')
COMMAND: COMPOSER_HOME=/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home COMPOSER_ROOT_VERSION=9.3.0 composer create-project --no-ansi drupal/recommended-project testproject 9.3.0 -vvv --repository /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/test_repository/packages.json
OUTPUT:
ERROR: Running 2.1.9 (2021-10-05 09:47:38) with PHP 7.4.2 on Darwin / 18.7.0
Reading ./composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer.json)
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading config file ./composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer.json)
Checked CA file /private/etc/ssl/cert.pem: valid
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json)
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Creating a "drupal/recommended-project" project at "./testproject"
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json)
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/test_repository/packages.json
Reading file:///Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/test_repository/packages.json
Writing /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/cache/repo/file----Applications-MAMP-tmp-php-build-workspace-78ba9d21989111c8bc37fa31ec93105eMQDn7J-test-repository-packages.json/packages.json into cache
Installing drupal/recommended-project (9.3.0)
Executing command (CWD): rm -rf '/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/testproject'
  - Installing drupal/recommended-project (9.3.0): Symlinking from composer/Template/RecommendedProject
Created project in /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/testproject
Reading ./composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Template/RecommendedProject/composer.json)
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading config file ./composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Template/RecommendedProject/composer.json)
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/config.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/composer.json)
Loading config file /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Reading /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/auth.json
Loading composer repositories with package information
Reading file:///Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/vendor_packages/packages.json
Writing /Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer-home/cache/repo/file----Applications-MAMP-tmp-php-build-workspace-78ba9d21989111c8bc37fa31ec93105eMQDn7J-vendor-packages-packages.json/packages.json into cache
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/core/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Plugin/Scaffold/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Plugin/ProjectMessage/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Plugin/VendorHardening/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Metapackage/PinnedDevDependencies/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Metapackage/CoreRecommended/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Executing command (/Applications/MAMP/tmp/php/build_workspace_78ba9d21989111c8bc37fa31ec93105eMQDn7J/composer/Metapackage/DevDependencies/): git rev-parse HEAD
Executing command (CWD): git rev-parse HEAD
Updating dependencies
Generating rules
Resolving dependencies through SAT
Looking at all rules.

Dependency resolution completed in 0.001 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires drupal/core-recommended ^9.3 -> satisfiable by drupal/core-recommended[9.3.0].
    - drupal/core-recommended 9.3.0 requires composer/semver 3.2.5 -> found composer/semver[3.2.4] but it does not match the constraint.
  Problem 2
    - Root composer.json requires drupal/core-dev ^9.3 -> satisfiable by drupal/core-dev[9.3.0].
    - drupal/core-dev 9.3.0 requires friends-of-behat/mink-browserkit-driver ^1.4 -> could not be found in any version, there may be a typo in the package name.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.


Failed asserting that 2 matches expected 0.

/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/Framework/Constraint/Equality/IsEqual.php:96
/Users/joachim/Sites/drupal-core-composer/repos/drupal/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php:300
/Users/joachim/Sites/drupal-core-composer/repos/drupal/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php:287
/Users/joachim/Sites/drupal-core-composer/repos/drupal/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php:209
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/Framework/TestResult.php:722
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:677
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:677
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:667
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/TextUI/Command.php:143
/Users/joachim/Sites/drupal-core-composer/vendor/phpunit/phpunit/src/TextUI/Command.php:96
```

**Comment by bbrala** [comment]
Hmm, i wanted to have a look, but i get another error. tried composer 1 to no avail. 

Unfortunately need to leave for home for now. 

```
vendor/bin/phpunit web/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php --filter=testTemplateCreateProject@recommended-project --debug
PHPUnit 9.5.12 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

Testing Drupal\BuildTests\Composer\Template\ComposerProjectTemplatesTest
Test 'Drupal\BuildTests\Composer\Template\ComposerProjectTemplatesTest::testTemplateCreateProject with data set "recommended-project" ('drupal/recommended-project', 'composer/Template/RecommendedProject', '/web')' started
Test 'Drupal\BuildTests\Composer\Template\ComposerProjectTemplatesTest::testTemplateCreateProject with data set "recommended-project" ('drupal/recommended-project', 'composer/Template/RecommendedProject', '/web')' ended


Time: 00:02.739, Memory: 14.00 MB

There was 1 error:

1) Drupal\BuildTests\Composer\Template\ComposerProjectTemplatesTest::testTemplateCreateProject with data set "recommended-project" ('drupal/recommended-project', 'composer/Template/RecommendedProject', '/web')
Error: Class "Drupal\Composer\Composer" not found

/home/localcopy/dcore/repos/drupal/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php:164
/home/localcopy/dcore/vendor/phpunit/phpunit/src/Framework/TestResult.php:726
/home/localcopy/dcore/vendor/phpunit/phpunit/src/Framework/TestSuite.php:677
/home/localcopy/dcore/vendor/phpunit/phpunit/src/Framework/TestSuite.php:677
/home/localcopy/dcore/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:670
/home/localcopy/dcore/vendor/phpunit/phpunit/src/TextUI/Command.php:143
/home/localcopy/dcore/vendor/phpunit/phpunit/src/TextUI/Command.php:96
```

**Comment by joachim-n** [comment]
Ok I think I sort of get it.

The test tries to install one of the Composer project templates that are in Drupal core. It does this by making a subdirectory, and installing the project into it -- with lots of spooky magic for Composer beforehand.

In the Composer project created with THIS template, the requirement for composer/semver comes:

> drupal/core                9.4.x-dev  requires  composer/semver (^3.0)

This loose requirement means that in my project, I have the latest version of composer/semver, 3.2.7.

But the test does an installation of the drupal/recommended-project template, and that requires drupal/core-recommended, which is a metapackage which fixes version numbers and has in particular:

>         "composer/semver": "3.2.6",

In the test, packagist is disabled, so that nothing can be downloaded, and the packages installed IN THE PROJECT are made available to Composer to install into the test project.

This means that the only version of composer/semver available within the test is the one installed in the actual project, 3.2.7, and the template being tested wants something else.

I am not sure how we fix. I am also confused how this test manages to run with a git clone of core -- will investigate.

**Comment by joachim-n** [comment]
Ah right, if you clone Drupal core and run the tests on that, you've installed Composer packages based on the composer.lock file which is in the core git clone, and that fixes the version:

```
            "name": "composer/semver",
            "version": "3.2.6",
```

**Comment by joachim-n** [comment]
So I guess the problem boils down to this: when you install packages in this template's project, Composer is free to install newer versions that what is declare in core's composer.lock and drupal/core-recommended.

The fix is to make this template respect those version requirements.

**Comment by joachim-n** [comment]
I think I've cracked it -- add the path to drupal/core-recommended in the repos/drupal as a repository, and require drupal/core-recommended with a wildcard version constraint. Needs some more testing!

**Comment by joachim-n** [comment]
Ok, that fixes the composer/semver problem, but not friends-of-behat/mink-browserkit-driver.

**Comment by joachim-n** [comment]
Fixed in 2332204bf2486712bce0869f7ef608b50784d840.
